### ISSUE TITLE
Re-compress Some Songs with UTF-8 File Name Encoding

### ISSUE THREAD

**Issue by WGzeyu** [beatmaps-io/beatsaver-main#201]
Some songs on Beat Saver are compressed without UTF-8 encoding for their file names. This can cause issues, as players with different regional settings may experience filename corruption if the song's file names contain local language characters. Consequently, only players with the same regional settings can access and play these songs. When players with different regional settings try to unpack these songs, the filenames may become garbled, and the game may fail to open the song due to the inability to locate the audio files.

I recommend implementing a script to inspect the zip files on the CDN. For compressed packages that do not use UTF-8 encoding for file names and contain audio file names with non-ASCII characters, the script should attempt to unpack the files using different regional encodings until the file names match those in the song's info (or directly rename them based on the info). Afterward, re-compress the files using UTF-8 encoding for file names. This process will not alter the song hash.

(Translate using ChatGPT)

[https://beatsaver.com/maps/252d6](https://beatsaver.com/maps/252d6)

![QQ截图20230901191357](https://github.com/beatmaps-io/beatsaver-main/assets/20312621/a15aee73-c5c4-445c-a1d2-1ab072fddfe1)
![QQ截图20230901191406](https://github.com/beatmaps-io/beatsaver-main/assets/20312621/8e3f3752-b9db-4ffb-ab76-44c9b554f6ec)
![QQ截图20230901191414](https://github.com/beatmaps-io/beatsaver-main/assets/20312621/ccb8e879-b2dd-4aaa-915c-07c13de80c6a)
![QQ截图20230901191423](https://github.com/beatmaps-io/beatsaver-main/assets/20312621/0fc0773d-59f7-4337-94e6-c3c4cda83722)
![QQ截图20230901191438](https://github.com/beatmaps-io/beatsaver-main/assets/20312621/d5103804-c0f8-4ec9-8725-9e6dcf7b587c)
![QQ截图20230901191450](https://github.com/beatmaps-io/beatsaver-main/assets/20312621/4da14fb0-e742-4d3e-8ee7-930bfc22ccec)

**Comment by Top-Cat** [comment]
I don't know if this was missed in translation or not so forgive me if I'm explaining something you already know.

It seems like there's a "bug" in the zip4j library beatsaver uses. We try to be minimally invasive so we rename the files rather than creating a new zip. In the renaming code by default the filenames are encoded as UTF-8 but the UTF-8 flag (bit 11 in the header) is unchanged. If the old filename was encoded with Cp437 then the newly encoded UTF-8 string will be decoded badly as shown in this issue.

To summarise the filenames above **ARE** UTF-8 encoded but the headers are wrong.

There doesn't seem to be an easy way in the library to change the flag but I think it should be sufficient to just maintain the source encoding when renaming files. This approach has been implemented in beatmaps-io/beatsaver-common-mp#17

Leaving this open as this only currently fixes new maps

**Comment by Top-Cat** [comment]
There are some more updates to come for new uploads but I've hopefully fixed ~216 existing maps. Shout if you find any missed examples and I'll investigate.
